controllers/watcher_controller.go

```diff
@@ -69,35 +71,45 @@ func (r *WatcherReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 	watcherObj := &v1alpha1.Watcher{}
 	if err := r.Get(ctx, client.ObjectKey{Name: req.Name, Namespace: req.Namespace}, watcherObj); err != nil {
 		logger.Error(err, "Failed to get reconciliation object")
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+		return ctrl.Result{Requeue: true}, client.IgnoreNotFound(err)
 	}
 
 	// check if deletionTimestamp is set, retry until it gets fully deleted
 	if !watcherObj.DeletionTimestamp.IsZero() {
 		err := r.IstioClient.RemoveVirtualServiceConfigForCR(ctx, req.NamespacedName)
 		if err != nil {
-			return ctrl.Result{Requeue: true}, err
+			return ctrl.Result{}, err
 		}
 		controllerutil.RemoveFinalizer(watcherObj, watcherFinalizer)
 		err = r.Update(ctx, watcherObj)
 		if err != nil {
-			return ctrl.Result{Requeue: true}, err
+			return ctrl.Result{}, err
 		}
-		return ctrl.Result{}, nil
+		return ctrl.Result{Requeue: true}, nil
 	}
 
 	// check finalizer on native object
 	if !controllerutil.ContainsFinalizer(watcherObj, watcherFinalizer) {
 		controllerutil.AddFinalizer(watcherObj, watcherFinalizer)
 		err := r.Update(ctx, watcherObj)
 		if err != nil {
-			return ctrl.Result{Requeue: true}, err
+			return ctrl.Result{}, err
 		}
-		return ctrl.Result{}, nil
+		return ctrl.Result{Requeue: true}, nil
 	}
-	err := r.IstioClient.UpdateVirtualServiceConfig(ctx, watcherObj)
+	virtualService, err := r.IstioClient.GetVirtualService(ctx)
+	if apierrors.IsNotFound(err) {
+		if _, err := r.IstioClient.CreateVirtualService(ctx, watcherObj); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to create virtual service %w", err)
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	err = r.IstioClient.UpdateVirtualServiceConfig(ctx, watcherObj, virtualService)
```

Comment:
can we encapsulate Create/Update into a single serverside apply instead of using the Istio Client?